### PR TITLE
#169590224 Builds a Central errorHandler

### PR DESCRIPTION
**Pull request by b0nbon1**
#### What does this PR do?
 sets up a central error handler that will handle all errors in the app.
#### Description of Task to be completed?
- add error handler file
- add 404 error handler
#### How should this be manually tested?
- clone the repository
- run `npm install` to install dependencies
- run the app using `npm run watch`
- test `/err` route on postman
#### Any background context you want to provide?
- Logger PR should be merged for this to work
#### What are the relevant pivotal tracker stories?
- [#169590224](https://www.pivotaltracker.com/n/projects/2407417/stories/169590224)
#### Screenshots:
image

#### Questions:
-  N/A

### CODE REVIEW

**Review by Mastel22 — COMMENTED** [review]
Kindly resolve for  Travis  first

**Review by Mastel22 — DISMISSED** [review]

**Review by gildniy — DISMISSED** [review]
Great.

File: src/utils/error.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+     * @param {string} message - The message to be passed for the error.
+     */
+  constructor(statusCode, message) {
+    super();
+    this.statusCode = statusCode;
+    this.message = message;
+  }
+}
+
+const handleError = (err, res) => {
```
Comment (timolinn) [inline src/utils/error.js]: This looks good, but I think this method tightly coupled to express's `Response` object ie. (it depends on `res` to work). How would you typically handle errors that are outside the context of `express`?

Comment (b0nbon1) [reply src/utils/error.js]: About the comment so I should remove the status and the res  so as to handle all errors generally??

Comment (b0nbon1) [reply src/utils/error.js]: @timolinn I implemented that it throws a 500 error and logs the error stack to the logger file is that okay??

**Review by gildniy — DISMISSED** [review]
Tested and looks good to me!

File: src/index.js [hunk 2]
```diff
@@ -22,41 +23,24 @@ if (!isProduction) {
...
-    });
-  });
-}
+app.use(router);
+app.use((req, res, next) => res.status(404).json({
+  error: 'Not found'
+}));
 
 
 app.use((err, req, res, next) => {
```
Comment (timolinn) [inline src/index.js:51]: This is nice, this means all errors would pass through express middleware. Kindly have a look at this implementation https://github.com/timolinn/node-starter/blob/13a9dfffba5d0565abdf91c55a44de43be90e259/src/app.js#L45.
There are two error handling middlewares there, for `production` and `development`.

**Review by gildniy — APPROVED** [review]
The way you set different responses, for prod/dev is a great idea.
Look good for me!

**Review by gildniy — DISMISSED** [review]

File: src/index.js [hunk 3]
```diff
@@ -15,48 +17,44 @@ app.use(cors());
...
-    errors: {
-      message: err.message,
-      error: {}
-    }
+  logger.error(`${err.status || 500} - ${err.message} - ${req.originalUrl} - ${req.method} - ${
+    req.ip
+  } - Stack: ${err.stack}`);
+
+  return res.status(err.statusCode).json({
+    message: 'Oops something went wrong'
```
Comment (timolinn) [inline src/index.js]: can we change this to `message: err.message`?

Comment (b0nbon1) [reply src/index.js]: ohh yeah yeah

**Review by gildniy — APPROVED** [review]
Working like charm.

**Review by Mastel22 — APPROVED** [review]
Tested and I can see it's working.Good Job

**Review by timolinn — APPROVED** [review]

**Pull request merged**